contracts/DissentVoting.sol

```diff
@@ -0,0 +1,388 @@
+    * @param _executesIfDecided Whether to also immediately execute newly created vote if decided
+    * @return voteId id for newly created vote
+    */
+    function newVote(bytes _executionScript, string _metadata, bool _castVote, bool _executesIfDecided)
+        external
+        auth(CREATE_VOTES_ROLE)
+        returns (uint256 voteId)
+    {
+        return _newVote(_executionScript, _metadata, _castVote, _executesIfDecided);
+    }
+
+    /**
+    * @notice Vote `_supports ? 'yea' : 'nay'` in vote #`_voteId`
+    * @dev Initialization check is implicitly provided by `voteExists()` as new votes can only be
+    *      created via `newVote(),` which requires initialization
+    * @param _voteId Id for vote
+    * @param _supports Whether voter supports the vote
+    * @param _executesIfDecided Whether the vote should execute its action if it becomes decided
+    */
+    function vote(uint256 _voteId, bool _supports, bool _executesIfDecided) external voteExists(_voteId) {
+        require(canVote(_voteId, msg.sender), ERROR_CAN_NOT_VOTE);
+        _vote(_voteId, _supports, msg.sender, _executesIfDecided);
+    }
+
+    /**
+    * @notice Execute vote #`_voteId`
+    * @dev Initialization check is implicitly provided by `voteExists()` as new votes can only be
+    *      created via `newVote(),` which requires initialization
+    * @param _voteId Id for vote
+    */
+    function executeVote(uint256 _voteId) external voteExists(_voteId) {
+        require(canExecute(_voteId), ERROR_CAN_NOT_EXECUTE);
+        _executeVote(_voteId);
+    }
+
+
+    /**
+    * @notice ACLOracle that indicates if `who` can perform certain action based on most recent yes vote
+    * @dev IACLOracle interface conformance
+    */    
+    function canPerform(address who, address where, bytes32 what, uint256[] how) external view returns (bool) {
+        return lastYeaVote[who] < getTimestamp64().sub(dissentWindow);
+    }
+
+    function isForwarder() public pure returns (bool) {
+        return true;
+    }
+
+    /**
+    * @notice Creates a vote to execute the desired action, and casts a support vote if possible
+    * @dev IForwarder interface conformance
+    * @param _evmScript Start vote with script
+    */
+    function forward(bytes _evmScript) public {
+        require(canForward(msg.sender, _evmScript), ERROR_CAN_NOT_FORWARD);
+        _newVote(_evmScript, "", true, true);
+    }
+
+    function canForward(address _sender, bytes) public view returns (bool) {
+        // Note that `canPerform()` implicitly does an initialization check itself
+        return canPerform(_sender, CREATE_VOTES_ROLE, arr());
+    }
+
+    function canVote(uint256 _voteId, address _voter) public view voteExists(_voteId) returns (bool) {
+        Vote storage vote_ = votes[_voteId];
+
+        //Check if voters stake hasn't changed since vote snapshot
+        // for more info see: https://github.com/1Hive/dissent-voting-app/issues/2#issuecomment-498684472
```

Comment:
Perhaps outline the issue briefly here, something like:

"The DissentVoting app's token must be a non-transferrable membership token eg addresses have 0 or 1 tokens, so we check the voter still has membership."

@lkngtn Is this correct? Otherwise if it's assumed the vote token is non-transferrable but it's amount can be other than 0 or 1 and increase/decrease through other means, eg contributing to the org or redeeming then there's a chance voters won't be able to vote. 

If it is only membership tokens then this applies restrictions to functions where the DissentVoting permission is granted. Eg, where DissentVoting protected functions use `msg.sender` to determine the owner of some assets, then the senders address must be the same as the one owning the voting token, and that must be the address that is granted the DissentVoting ACLOracle permission to that function.

In the `redeem()` case where we don't use `msg.sender` but a signature to determine the redeemer, this means  someone who hasn't voted yes recently can redeem for someone who has voted yes recently, probably unlikely to happen but worth noting. 

Also, as it stands you can't give the DissentVoting ACLOracle permission for a function to a forwarder, eg a TokenManager, because it's very unlikely the forwarder address will ever vote and therefore users of the forwarder will always be able to call the function.

So AFAIK to limit access to the `redeem()` function would currently require adding individual permissions for each user that wants to call it.

To enable a forwarder to call a function with the ACLOracle permission I think you would have to pass the users signature through the auth modifier eg `authP(REDEEM_ROLE, arr(msgHash, signature, dissentWindow))`, get the address from the signature in the `DissentVoting` app and use that in the `canPerform()` function.

Of course this requires modification of any function wishing to use the `DissentVoting` ACLOracle. 

Curious to know if my claims above are accurate but this may require more thought.

Reply:
> "The DissentVoting app's token must be a non-transferrable membership token eg addresses have 0 or 1 tokens, so we check the voter still has membership."

> @lkngtn Is this correct? Otherwise if it's assumed the vote token is non-transferrable but it's amount can be other than 0 or 1 and increase/decrease through other means, eg contributing to the org or redeeming then there's a chance voters won't be able to vote.

Definitely should allow for token balances other than 0 or 1. It would likely be standard 18 decimals. AFAIK there shouldnt need to be any special handling regarding how the token is configured (it can be either transferrable or not not), but in practice the mechanism is not useful if the tokens are transferrable since they can avoid the permission restriction if they were to transfer their tokens to a different address and use that other address (which hasn't voted) to perform the action instead. 

> So AFAIK to limit access to the redeem() function would currently require adding individual permissions for each user that wants to call it.

Hmm this does seem like an issue. The intention was to have the redeem role granted to `ANY_ENTITY` with then use the ACL Oracle parameter to prevent people who have recently voted yes from successfully calling the function. I thought that this would work but because of the way Redemptions handles interactions with forwarders it may not. Specifically if the user routes their request through a forwarder (any forward doesn't even need to be installed in the org) they are able to bypass the restriction, since the actual request will be coming from a different address. So using ANY_ENTITY with this ACLOracle would not be effective, unless we adjust the redeem function to only allow direct interactions with msg.sender(). 

The simpler, though not very elegant, solution would be to add an additional redeem function and role, something like "direct redeem". This would allow an organization that wants to use this ACL oracle with `ANY_ENTITY` to use that and discard the normal redeem function in order to create a secure transaction path. (alternatively we could make the redeem function direct only like we had originally designed it and create a seperate forwarder redeem function). Ultimately, as you point out, it doesn't seem like it will work well if there are both forwarders and ACL oracles in the transaction path.